### PR TITLE
Add Visible modes function from Organisation to Users too

### PR DESCRIPTION
**Pull request by sergey-dryabzhinsky**
You can limit or hide organisations. This pull make it also posible for users

- new strings to translte
- add checkbox to user profile form
- add checkbox to admin user.edit form
- filter explore page user search
- filter api admin and public user searches
- allow admins view "hidden" users

### CODE REVIEW

**Comment by sergey-dryabzhinsky** [comment]
Thats for https://github.com/go-gitea/gitea/issues/14477#issuecomment-854323537

File: templates/explore/users.tmpl
Comment (lunny) [inline templates/explore/users.tmpl]: We should filter them when Query but not in the UI. Assume all 10 users on this page chosed hide themselves? What will happen?

Comment (sergey-dryabzhinsky) [reply templates/explore/users.tmpl]: Oh, it's true. I'll update query then.

**Comment by 6543** [comment]
two things:
 * it would be nice if we have a test case of this in integration tests ...
 * we should hide them from https://try.gitea.io/api/swagger#/user/userSearch beside user itselfe ...

**Comment by sergey-dryabzhinsky** [comment]
@6543 
- I'll try to add some tests
- ok, if it public accessible

**Comment by sergey-dryabzhinsky** [comment]
@6543 But admin must be able to list them too.

File: models/user.go
Comment (6543) [inline models/user.go]:
```suggestion
```

File: models/user.go
Comment (6543) [inline models/user.go]:
```suggestion
	if opts.Actor != nil {
	   if !opts.Actor.IsAdmin() {
		cond = cond.And(builder.Eq{"hide_from_explore_page": false}.Or(builder.Eq{"id": opts.Actor.ID}))
		}
	} else {
		cond = cond.And(builder.Eq{"hide_from_explore_page": false})
	}
```

**Review by 6543 — COMMENTED** [review]
something like this ...

dont just apply since it's not proper formated - just my idears i had in mind ;)

File: models/user.go
Comment (sergey-dryabzhinsky) [reply models/user.go]: Added a bit formated, remove useles field in searchOpts

**Comment by sergey-dryabzhinsky** [comment]
Hmmm... I don't understand how to check is current request made by Admin User inside API call.

**Comment by sergey-dryabzhinsky** [comment]
I wander if token-auth is ever realy tested

**Comment by sergey-dryabzhinsky** [comment]
Done it not right. Admin users page - without hidden users.

**Comment by sergey-dryabzhinsky** [comment]
Or it's just api context missing User.

**Review by 6543 — APPROVED** [review]

**Comment by zeripath** [comment]
We'll need to fix up the language in this PR. Mergers please do not merge this without fixing the language.

---

I'm not certain really why we can't just use the Visibility field that's already present in the User table. 

We could use Public, Limited and Private like we have for Organizations with the following meanings: 

* Public - appears on the Explore page.
* Limited - appears on the Explore page only for logged in users.
* Private - never appears on the Explore page.

Equivalently, repositories for users would then gain the same visibilities on the Explore Repos pages.

**Review by 6543 — CHANGES_REQUESTED** [review]
as per zeripath

**Comment by 6543** [comment]
@zeripath you are right, totaly forgot we already have a field for that ...

**Comment by sergey-dryabzhinsky** [comment]
@zeripath English is not my main language, so suggestion are welcome.

About `visible` suggestion: the meaning of this PR is about `Public` or `Private` only selection.
`Limited` never meant to be here as it pointless as I think.
Well it may have point if Gitea is used by one Team/Organization for all repos.

The next level PR will be `Only to my Teams`.

**Comment by sergey-dryabzhinsky** [comment]
Ok. I dived deeper into the code.
It seems that this property is also suitable for limiting to one command.

Please wait.

**Comment by sergey-dryabzhinsky** [comment]
Do we need `DEFAULT_USER_VISIBILITY` config option? Or they alway be `public`?

**Comment by 6543** [comment]
I think we can leave default to public, ... for setting page just replace the checkbox with a drop-down or do it similar to organisation settings for the sake of consistency ...

**Comment by sergey-dryabzhinsky** [comment]
@6543 Already added option ;) Set to `public` by default.

Yes, I added html to user form templates as for orgs.
But it looks messy:

User profile settings.
Снимок экрана_2021-06-10_04-56-59

Admin user edit.
Снимок экрана_2021-06-10_04-58-16

Any suggestions?

**Comment by 6543** [comment]
change both to a dropdown solution? - drop down is clean & strait forward

**Comment by sergey-dryabzhinsky** [comment]
Like that?
Снимок экрана_2021-06-10_05-54-55

**Comment by sergey-dryabzhinsky** [comment]
User profile edit:
Снимок экрана_2021-06-10_06-28-15

**Comment by zeripath** [comment]
No worries about the translations - you write better English than I write Russian. 

I'd normally have just suggested improvements but thought that the other part of my comment would likely make them out of date. If we'd decided that that we shouldn't use the visibility field for this then I wanted to warn that we needed to fix the language before merging.

**Comment by sergey-dryabzhinsky** [comment]
@zeripath Ok.
We decided to USE Visibility field as it seems not used anywhere.
@6543 ?

**Comment by sergey-dryabzhinsky** [comment]
I don't thinj that api is good now.
Everywhere Visibility is string type, but for user - integer.
But if it string - User in releation can't be loaded via api.

**Comment by 6543** [comment]
@sergey-dryabzhinsky Visibility field is used by organisations (those are overloaded users, in terms of DB structure) ...

**Comment by sergey-dryabzhinsky** [comment]
Damn.
Return old behaviour?

File: integrations/api_user_search_test.go
Comment (6543) [inline integrations/api_user_search_test.go]:
```suggestion
		assert.EqualValues(t, user.Visibility, api.VisibleTypePrivate)
```

File: modules/structs/user.go
Comment (6543) [inline modules/structs/user.go]: can you change it to string?

File: modules/structs/org_type.go
Comment (6543) [inline modules/structs/org_type.go]: you dont have to since VisibleTypePublic is alias for int 0

File: modules/structs/user.go
Comment (sergey-dryabzhinsky) [reply modules/structs/user.go]: Done

File: modules/structs/org_type.go
Comment (sergey-dryabzhinsky) [reply modules/structs/org_type.go]: Removed

**Comment by sergey-dryabzhinsky** [comment]
Now API tests fails cos they suggests that Visibility from API response are integer.
```
api_repo_test.go:478: 
210        	Error Trace:	integration_test.go:512
211        	            				api_repo_test.go:478
212        	Error:      	Received unexpected error:
213        	            	models.Repository.Owner: models.User.Visibility: readUint64: unexpected character: �, error found in #10 byte of ...|ibility":"public","u|..., bigger context ...|n":"","website":"","description":"","visibility":"public","username":"user1"},"name":"moveME","full_|...
214        	Test:       	TestAPIRepoTransfer
```

**Comment by sergey-dryabzhinsky** [comment]
Seems like there is 3 ways available:
- make User.Visibility of type `string`, all the way through `gitea`, like `EmailNotificationsPreference`
- make apiUser.Visibility of type `int64` again, to make relations work
- make new flag User.HideFromExplorePage just to hide from `explore/users`

File: modules/structs/user.go
Comment (6543) [reply modules/structs/user.go]: @sergey-dryabzhinsky thanks

**Comment by 6543** [comment]
I'll have a look :)

**Review by 6543 — APPROVED** [review]

**Comment by 6543** [comment]
@sergey-dryabzhinsky should be ready - can you test if & tell me if you are ok with my changes too :D

File: custom/conf/app.example.ini
Comment (zeripath) [inline custom/conf/app.example.ini]: I think the text above here needs to be changed.

File: options/locale/locale_en-US.ini
Comment (zeripath) [inline options/locale/locale_en-US.ini:733]: hmm... is this true? Can you still see public repositories of private users?

**Comment by sergey-dryabzhinsky** [comment]
@6543 Thanks! I'll test that changes ASAP.

File: options/locale/locale_en-US.ini
Comment (sergey-dryabzhinsky) [reply options/locale/locale_en-US.ini:733]: Yes, repo Visibility logic not changed.

**Comment by sergey-dryabzhinsky** [comment]
@6543 All looks good to me.

**Comment by 6543** [comment]
hmm coming this far we can make org Visible and user Visible feature parity ...
... this would simplify some things

throu I think it will miss the aim you had in mid @sergey-dryabzhinsky ?

we just have to remove/add some checks specific to user.type = organisaton ...

**Comment by sergey-dryabzhinsky** [comment]
@6543 Yeah, I prefer to keep things simple. There is enough complexity already.

**Comment by sergey-dryabzhinsky** [comment]
@6543 I mean I would like to leave this code in this state.

Do you have something greater in mind and based on this code?
I think it should be merged first or backed for future.

File: templates/swagger/v1_json.tmpl
Comment (noerw) [inline templates/swagger/v1_json.tmpl]: Is public actually the default? If defatul is dynamic (my understanding) I'd only put
```suggestion
          "description": "User visibility level option: public, limited, private",
```

File: templates/user/settings/profile.tmpl
Comment (noerw) [inline templates/user/settings/profile.tmpl]: `value="{{.visibility}}"` always renders to `value=""` for me. Not sure why it works in the admin panel but not here, as I don't understand where the `.visibility` var should be set

**Review by noerw — CHANGES_REQUESTED** [review]
Thanks, great feature  :) Found some issues though (as per comments).

Also a suggestion: Shouldn't users be able to select `default (server-wide preference: $DEFAULT_USER_VISIBILITY)` from the dropdowns too?

File: templates/swagger/v1_json.tmpl
Comment (sergey-dryabzhinsky) [reply templates/swagger/v1_json.tmpl]: Probably yes, do not need to mention it as default.

File: templates/user/settings/profile.tmpl
Comment (sergey-dryabzhinsky) [reply templates/user/settings/profile.tmpl]: Yes, it's should be something like `.SignedUser.Visibility`

**Comment by sergey-dryabzhinsky** [comment]
@noerw 
> Also a suggestion: Shouldn't users be able to select default (server-wide preference: $DEFAULT_USER_VISIBILITY) from the dropdowns too?

You mean to mark app-default level in selects?

**Comment by noerw** [comment]
@sergey-dryabzhinsky The fix for the empty dropdown didn't work for me. Is the dropdown populated on your machine? 🤔

> You mean to mark app-default level in selects?
> Or add special option "default"?

I mean the latter. Though looking at how you implemented it, this wouldn't be a trivial change, so I'm fine with leaving it as is.

**Comment by 6543** [comment]
I to what i have in mind: make it feature complete to how it works with orgs

**Comment by 6543** [comment]
So there are in theory just some IsOrg checks to be removed that handle the user has permission to see repo

**Comment by sergey-dryabzhinsky** [comment]
@6543 I have doubts that it is simple as it sounds

**Comment by sergey-dryabzhinsky** [comment]
Found new bug: if you make organization non-public - then you can't see it in admin panel - admin/orgs. Even if you admin.

**Comment by 6543** [comment]
> @6543 I have doubts that it is simple as it sounds

done

**Review by 6543 — APPROVED** [review]

**Comment by sergey-dryabzhinsky** [comment]
@6543 I think this issue need to be renamed corresponding to new changes.

**Review by noerw — CHANGES_REQUESTED** [review]
The API still exposes private users (it always did wrt orgs? :grimacing:):
- http://localhost:3000/api/v1/users/norwin
- http://localhost:3000/api/v1/orgs/foobar

Otherwise LGTM

**Comment by sergey-dryabzhinsky** [comment]
@noerw By new code idea it should not.

File: models/user.go
Comment (sergey-dryabzhinsky) [inline models/user.go]: Is it legitimate? Isn't here org membership must be checked?

File: routers/api/v1/user/user.go
Comment (noerw) [inline routers/api/v1/user/user.go]: This must run after the `ctx.Written()` check below, otherwise we generate an internal server error.

Also there still is an information leak because the 404 response differs between a "true" 404, and this case.
Try the following `.patch`:
```patch
diff --git a/routers/api/v1/user/helper.go b/routers/api/v1/user/helper.go
index fcdac257e..a3500e0ee 100644
--- a/routers/api/v1/user/helper.go
+++ b/routers/api/v1/user/helper.go
@@ -17,7 +17,7 @@ func GetUserByParamsName(ctx *context.APIContext, name string) *models.User {
 	user, err := models.GetUserByName(username)
 	if err != nil {
 		if models.IsErrUserNotExist(err) {
-			if redirectUserID, err := models.LookupUserRedirect(username); err == nil {
+			if redirectUserID, err2 := models.LookupUserRedirect(username); err2 == nil {
 				context.RedirectToUser(ctx.Context, username, redirectUserID)
 			} else {
 				ctx.NotFound("GetUserByName", err)
diff --git a/routers/api/v1/user/user.go b/routers/api/v1/user/user.go
index 44d392c47..1cf77bea4 100644
--- a/routers/api/v1/user/user.go
+++ b/routers/api/v1/user/user.go
@@ -103,11 +103,12 @@ func GetInfo(ctx *context.APIContext) {
 	//     "$ref": "#/responses/notFound"
 
 	u := GetUserByParams(ctx)
-	if !u.IsVisibleToUser(ctx.User) {
-		ctx.JSON(http.StatusNotFound, nil)
+	if ctx.Written() {
 		return
 	}
-	if ctx.Written() {
+	if !u.IsVisibleToUser(ctx.User) {
+		// fake ErrUserNotExist error message to not leak information about existence
+		ctx.NotFound("GetUserByName", models.ErrUserNotExist{Name: ctx.Params(":username")})
 		return
 	}
 
```

File: models/user.go
Comment (6543) [reply models/user.go]: Its about seeing a private user... it whuld be hidden completly, but follow one so he can see you is a nice idear isnt it?

Comment (sergey-dryabzhinsky) [reply models/user.go]: Still not enough. Tried to roundup that.

File: routers/api/v1/user/user.go
Comment (sergey-dryabzhinsky) [reply routers/api/v1/user/user.go]: Done

**Review by noerw — COMMENTED** [review]
functionality-wise I'm good with this now, good work!

It's just that I think we shouldn't have both `models.HasOrgOrUserVisible()` and `models.IsVisibleToUser()`: can't those functions be merged, maybe into two codepaths for orgs and users?

**Comment by zeripath** [comment]
> IsVisibleToUser

Yes these functions appear very similar but slightly different and will be a source of confusion.

---

I should also add that permissions tests should usually be subject before object, as in *Alice* can see *Bob*, not *Bob* is visible to *Alice* - I appreciate that this confusion is already present so it's just a point to remind me that when we do the permissions refactor that we will need to address this (amongst other things)

**Comment by zeripath** [comment]
OK I think the IsVisibleTo function could be improved to:

```go
		// Now we need to check if they in some team together
		count, err := x.Table("team_user").
			Where(
				builder.And(
					builder.Eq{"uid": viewer.ID},
					builder.Or(
						builder.Eq{"org_id": u.ID},
						builder.In("org_id",
							builder.Select("org_id").
								From("team_user", "t2").
								Where(builder.Eq{"uid": u.ID}))))).
			Count(new(TeamUser))
		if err != nil {
			return false
		}

		if count < 0 {
			// No common teams or organizations
			return false
		}
```

**Comment by zeripath** [comment]
Then this will work for both organizations and users - and when we start allowing users to have teams this will still work.

**Comment by 6543** [comment]
I dont think users will need teams ... organisations are there for but ... yes it would work

**Comment by zeripath** [comment]
agh hang on a sec ... 

What's the test supposed to be? Are they supposed to be in team or an organization together?

I the current implementation (and my suggestion above) are doing organization not team.

I would have thought it should be team.

File: models/repo_permission.go
Comment (zeripath) [inline models/repo_permission.go]:
```suggestion
	// Allow user if they are collaborator of a repo within a private user or a private organization but not a member of the organization itself
```

**Review by zeripath — APPROVED** [review]

**Pull request merged**
